### PR TITLE
Feat:Add Pinecone as a new vector database

### LINKED ISSUE #8892: Add Pinecone as a new vector database
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing? Tell me about your story.

Add Pinecone as a new vector database

### 2. Additional context or comments

_No response_

### 3. Can you help us with this feature?

- [X] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by Howe829**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

How to use:
PS:You need to get the <PINECONE_API_KEY> and create a pinecone Index from the pinecone console.

First,  add <PINECONE_API_KEY>, <PINECONE_INDEX>, <PINECONE_INDEX_DIMENSION> to your .env file.
Second, change the <VECTOR_STORE> to 'pinecone'.

Fixes #8892 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: api/tests/integration_tests/vdb/pinecone/test_pinecone.py
Comment (yaoice) [inline api/tests/integration_tests/vdb/pinecone/test_pinecone.py:1]: please add ut in `dev/pytest/pytest_vdb.sh` to execute.

**Review by yaoice — CHANGES_REQUESTED** [review]
See comments

**Comment by Howe829** [comment]
> See comments

Hi! I'm glad to see your comment, I've resolved it.PTAL.

**Comment by crazywoola** [comment]
@Howe829 Hello, please resolve deps issue here.

**Comment by Howe829** [comment]
> @Howe829 Hello, please resolve deps issue here.

Hi there! I've resolved the deps issues.The Pytest failed the due to the api key  provided in test_pinecone.py is not real.

**Comment by yaoice** [comment]
> > @Howe829 Hello, please resolve deps issue here.
> 
> Hi there! I've resolved the deps issues.The Pytest failed due to the api key provided in test_pinecone.py is not real.

I think you need to mock api with calling pinecone to run pytest successfully.

**Comment by Howe829** [comment]
> > > @Howe829 Hello, please resolve deps issue here.
> > 
> > 
> > Hi there! I've resolved the deps issues.The Pytest failed due to the api key provided in test_pinecone.py is not real.
> 
> I think you need to mock api with calling pinecone to run pytest successfully.

Thanks for your advice, all checks have passed.

**Comment by crazywoola** [comment]
Seems there more conflicts here.

**Comment by Howe829** [comment]
> Seems there more conflicts here.

Resolved.PTAL @crazywoola

**Comment by crazywoola** [comment]
@Howe829 Hello there are some errors in the tests.

File: api/tests/integration_tests/vdb/pinecone/test_pinecone.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    added_doc_ids = test_instance.add_texts()
+    test_instance.delete_by_ids(added_doc_ids)
+    test_instance.delete_vector()
+
+
+@pytest.fixture
+def pinecone_mock(test_instance):
+    mock = MagicMock()
+    mock.has_index.return_value = True
+    mock.Index.return_value.query.return_value = {
```
Comment (yaoice) [inline api/tests/integration_tests/vdb/pinecone/test_pinecone.py:48]: I think it is more reasonable to mock the methods of pindex.delete, query, and upsert.

**Pull request closed without merging**